### ISSUE TITLE
AutoML - different results for the same input data

### ISSUE THREAD

**Issue by kicaj29** [dotnet/machinelearning-samples#1011]
Hi

I am playing with AutoML for MultiClassification.
My csv file has 2 columns: Words and ClassId. It has also 15 rows, first 5 rows have classId=10, next 5 have classId=11 and last 5  have classId = 12.

I create `TrainTestData` in this way:

```
TrainTestData trainValidationData = ctx.Data.TrainTestSplit(data, testFraction: 0.2);
```

and I run AutoML in this way:

```
SweepablePipeline pipeline =
         ctx.Transforms.Text.FeaturizeText(inputColumnName: @"Words", outputColumnName: @"Words")
        .Append(ctx.Transforms.Concatenate(@"Features", new[] { @"Words" }))
        .Append(ctx.Transforms.Conversion.MapValueToKey(outputColumnName: @"ClassId", inputColumnName: @"ClassId", addKeyValueAnnotationsAsText: false))
        .Append(ctx.Auto().MultiClassification(labelColumnName: columnInference.ColumnInformation.LabelColumnName));

AutoMLExperiment experiment = ctx.Auto().CreateExperiment();

experiment
    .SetPipeline(pipeline)
    .SetMulticlassClassificationMetric(MulticlassClassificationMetric.MacroAccuracy, labelColumn: columnInference.ColumnInformation.LabelColumnName)
    .SetTrainingTimeInSeconds(60)
    .SetDataset(trainValidationData);

var monitor = new AutoMLMonitor(pipeline);
experiment.SetMonitor(monitor);

TrialResult experimentResults = await experiment.RunAsync();
```

Can you explain why multiple runs returns different results? For example one run can generate model with `MacroAccuracy=0` and second run can generate model with `MacroAccuracy=1`, or `MacroAccuracy=0.5`.

The same behavior I observed when I was using designer in VS - also runs for the same input data and params generate results with different `MacroAccuracy`.

Basically how I can relay on AutoML when it looks it generate kind of random results for the same input and parameter of run?

**Comment by superichmann** [comment]
yo yo, I am just a user like you but I came across this so why not ;;]

when u initialize ctx ensure you set a seed on `public MLContext (int? seed = default);` which will enable deterministic reproducable results as outlined [here](https://learn.microsoft.com/en-us/dotnet/api/microsoft.ml.mlcontext.-ctor?view=ml-dotnet#microsoft-ml-mlcontext-ctor(system-nullable((system-int32)))).

as well make sure you send a seed to [TrainTestSplit ](https://learn.microsoft.com/en-us/dotnet/api/microsoft.ml.dataoperationscatalog.traintestsplit?view=ml-dotnet).

send same seed all the time and then you will get the same results.

As well set [MaxModels](https://github.com/dotnet/machinelearning/pull/5471) instead of `.SetTrainingTimeInSeconds(60)` to ensure no external programs hog the cpu and give you different results for same time period..

**Comment by kicaj29** [comment]
this is really comprehensive answer, thanks !
